text/0000-safer-transmute.md

```diff
@@ -643,93 +532,73 @@ Even with `NeglectValidity`, the compiler will statically reject transmutations
-
-### Implementing `Constructible`
-The compiler implements `Constructible<Scope>` for `T` if `T` is fully implicitly constructible in the scope uniquely identified by the type `Scope`.
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
 
-A type `T` is fully implicitly constructible in a particular scope if:
-  - `T`'s implicit constructor is reachable from the scope, and either:
-    - `T` has no fields, or
-    - `T`'s fields are fully implicitly constructible from the scope.
+## Implementation Guidance
+Two items in this RFC require special compiler support:
+  - `Muckable`
+  - `TransmuteFrom`
+  - `differing_sizes` lint
 
-The `Constructible` trait does not ever need to be made stable, or even visible (à la the virtual `Freeze` trait). It is merely a useful device for implementing `TransmuteFrom`.
+### Implementing `Muckable`
+The `Muckable` marker trait is similar to `Copy` in that all fields of a `Muckable` type must, themselves, be `Muckable`.
 
 ### Implementing `TransmuteFrom`
 The implementation of `TransmuteFrom` is completely internal to the compiler (à la [`Sized`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyS.html#method.is_sized) and [`Freeze`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyS.html#method.is_freeze)).
 
 #### Constructability and Transmutation
-Unless `NeglectConstructability` is used as `Neglect` option, a `Src` is *safely* transmutable into `Dst` in a given `Scope` if:
-  1. `Dst: Constructible<Scope>`
-  2. `Dst: TransmuteFrom<Src, Scope, Neglect>`
+A `Src` is *safely* transmutable into `Dst` in a given if:
+  1. `Dst: Muckable`
+  2. `Dst: TransmuteFrom<Src, Neglect>`
+  3. `NeglectSafety` ∉ `Neglect`
 
 If `Src` is a mutatable reference, then additionally:
-  1. `Src: Constructible<Scope>`
+  1. `Src: Muckable`
 
-### Implementing `differing_sizes`
-The `differing_sizes` lint reports a compiler warning when the source and destination types of a `transmute!()`, `transmute_into` or `transmute_from` invocation differ. This lint shall be warn-by-default.
+### Implementing `differing_sizes` lint
+The `differing_sizes` lint reports a compiler warning when the source and destination types of a `transmute!()`, `transmute_from` or `unsafe_transmute_from` invocation differ. This lint shall be warn-by-default.
 
 ### Minimal Useful Stabilization Surface
 Stabilizing *only* this subset of the Initial Smart Implementation will cover many use-cases:
   - `transmute!()`
 
 To define traits that generically abstract over `TransmuteFrom`, these items must be stabilized:
-  - `Here!()`
   - `TransmuteFrom`
   - `TransmuteOptions` and `SafeTransmuteOptions`
```

Comment:
I believe this has been asked before, but it's still not clear to me. Why are `TransmuteOptions` necessary to stabilize `TransmuteFrom`? Is it bot possible to add the additional type parameter (defaulting to `()`) in the future?

Reply:
You *can* stabilize `TransmuteFrom` without stabilizing `TransmuteOptions` or `SafeTransmuteOptions`. These are only necessary to define traits that totally generically abstract over `TransmuteFrom`; e.g., the `Vec`/Slice-casting example.

Reply:
I'm wondering if that's worth being explicit about. The way it's worded currently is not super clear. We can say:

Stabilizing *only* this subset of the Initial Smart Implementation will cover many use-cases:
 - `transmute!()`

To define traits that generically abstract over sound, safe, and stable transmutation these items must be stabilized:
- `TransmuteFrom`

To define traits that totally generically abstract over `TransmuteFrom` including transmutations that are sound but not statically guaranteed to be safe and/or stable, these items must be stabilized:
  - `TransmuteOptions` and `SafeTransmuteOptions`